relay/writer/file.go

```diff
@@ -0,0 +1,381 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package writer
+
+import (
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser"
+	gmysql "github.com/siddontang/go-mysql/mysql"
+	"github.com/siddontang/go-mysql/replication"
+	"github.com/siddontang/go/sync2"
+
+	"github.com/pingcap/dm/pkg/binlog"
+	"github.com/pingcap/dm/pkg/binlog/common"
+	"github.com/pingcap/dm/pkg/binlog/event"
+	bw "github.com/pingcap/dm/pkg/binlog/writer"
+	"github.com/pingcap/dm/pkg/log"
+)
+
+// FileConfig is the configuration used by the FileWriter.
+type FileConfig struct {
+	RelayDir string // directory to store relay log files.
+	Filename string // the startup relay log filename, if not set then a fake RotateEvent must be the first event.
+}
+
+// FileWriter implements Writer interface.
+type FileWriter struct {
+	cfg *FileConfig
+
+	mu    sync.RWMutex
+	stage common.Stage
+
+	// underlying binlog writer,
+	// it will be created/started until needed.
+	out *bw.FileWriter
+
+	filename sync2.AtomicString // current binlog filename
+}
+
+// NewFileWriter creates a FileWriter instances.
+func NewFileWriter(cfg *FileConfig) Writer {
+	w := &FileWriter{
+		cfg: cfg,
+	}
+	w.filename.Set(cfg.Filename) // set the startup filename
+	return w
+}
+
+// Start implements Writer.Start.
+func (w *FileWriter) Start() error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	if w.stage != common.StageNew {
+		return errors.Errorf("stage %s, expect %s, already started", w.stage, common.StageNew)
+	}
+	w.stage = common.StagePrepared
+
+	return nil
+}
+
+// Close implements Writer.Close.
+func (w *FileWriter) Close() error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	if w.stage != common.StagePrepared {
+		return errors.Errorf("stage %s, expect %s, can not close", w.stage, common.StagePrepared)
+	}
+
+	var err error
+	if w.out != nil {
+		err = w.out.Close()
+	}
+
+	w.stage = common.StageClosed
+	return errors.Trace(err)
+}
+
+// Recover implements Writer.Recover.
+func (w *FileWriter) Recover(p *parser.Parser) (*RecoverResult, error) {
+	w.mu.RLock()
+	defer w.mu.RUnlock()
+
+	if w.stage != common.StagePrepared {
+		return nil, errors.Errorf("stage %s, expect %s, please start the writer first", w.stage, common.StagePrepared)
+	}
+
+	return w.doRecovering(p)
+}
+
+// WriteEvent implements Writer.WriteEvent.
+func (w *FileWriter) WriteEvent(ev *replication.BinlogEvent) (*Result, error) {
+	w.mu.RLock()
```

Comment:
`Lock` or `RLock`?

Reply:
The purpose of this mutex is to ensure only do the write operation in `StagePrepared`, so I use `RLock`. for `WriteEvent` itself is not supported by concurrent usage, and I added a comment in _relay/writer/write.go_ L62.

Reply:
changed to use `Mutex` instead of `RWMutex`